### ISSUE TITLE
Model is not learning at all

### ISSUE THREAD

**Issue by SaadSaeed150** [AI4Finance-LLC/FinRL-Library#8]
I used FinRL_single_stock_trading.ipynb to get the overview how to use the library but after training model and predicting the actions on trade data when you see the df_actions list the results are bit weird the agent is either selling all stocks(hmax) on each time stamp and then follow the same pattern throughout the trading data (selling all stocks at each time stamp) or sometimes on running it buys all stocks (hmax) on each time stamp and repeats the same action on every timestamp should i be worried about these results aren't they bit abnormal ?

I tried multiple trading data split variations (trading data : 2015-current ) (trading data: 2019-current) but same results 
I am using Apple stocks the one which was used in FinRL_single_stock_trading.ipynb and the model was TD3 
I tried other models as well still they were giving almost same result (at some timestamp the stock buy or sell were hmax - {1..5}) but the same results can you please verify these changes i think there is some issue in the repo code

**Comment by SaadSaeed150** [comment]
I have'nt changed any line in your jupyter file still the stocks are either all being sold or buyed there is no other action taken agent is taking after that i changed the data to IBM now the annual profit is -6.03% meaning the model is not learning at any point

**Comment by SaadSaeed150** [comment]
@BruceYanghy is there any chance some commit have made the issue btw thank you soo much for this amazing work and explanation of each and every step

**Comment by BruceYanghy** [comment]
For single stock trading, I posted the tensorboard episode reward chart and the loss function chart: https://towardsdatascience.com/finrl-for-quantitative-finance-tutorial-for-single-stock-trading-37d6d7c30aac. 

It's not that the model is not learning, the reward converges to a point, and the loss is decreasing, if you observe the backtesting results, it's basically a little bit better than the buy-and-hold strategy, it means that the current information is not enough, the model just gives the buy-and-hold strategy.

I am current trying to put Earnings report financial indicators and news data NLP sentiment into the model.

**Comment by BruceYanghy** [comment]
> For single stock trading, I posted the tensorboard episode reward chart and the loss function chart: https://towardsdatascience.com/finrl-for-quantitative-finance-tutorial-for-single-stock-trading-37d6d7c30aac.
> 
> It's not that the model is not learning, the reward converges to a point, and the loss is decreasing, if you observe the backtesting results, it's basically a little bit better than the buy-and-hold strategy, it means that the current information is not enough, the model just gives the buy-and-hold strategy.
> 
> I am current trying to put Earnings report financial indicators and news data NLP sentiment into the model.

**Comment by BruceYanghy** [comment]
> I used FinRL_single_stock_trading.ipynb to get the overview how to use the library but after training model and predicting the actions on trade data when you see the df_actions list the results are bit weird the agent is either selling all stocks(hmax) on each time stamp and then follow the same pattern throughout the trading data (selling all stocks at each time stamp) or sometimes on running it buys all stocks (hmax) on each time stamp and repeats the same action on every timestamp should i be worried about these results aren't they bit abnormal ?
> 
> I tried multiple trading data split variations (trading data : 2015-current ) (trading data: 2019-current) but same results
> I am using Apple stocks the one which was used in FinRL_single_stock_trading.ipynb and the model was TD3
> I tried other models as well still they were giving almost same result (at some timestamp the stock buy or sell were hmax - {1..5}) but the same results can you please verify these changes i think there is some issue in the repo code

You can also refer to this article: **https://arxiv.org/pdf/2004.06627.pdf**

**Comment by wangshizhang** [comment]
The same as your work, it seems its environment settings are not so suitable (even not so readable), maybe you can modify the environment.

**Comment by Bturan19** [comment]
I tried to create a lekage feature, to see if the model is able to exploit this feature and make some amazing profit. But even though, I leaked the future information as a feature, model still behaves like buy and hold.

**Comment by TekpreXyz** [comment]
Hello, i try to do your great tutorial:
https://towardsdatascience.com/finrl-for-quantitative-finance-tutorial-for-single-stock-trading-37d6d7c30aac
but in jupyter when i try to:
 !pip install tensorflow==1.15.4
i get
ERROR: Could not find a version that satisfies the requirement tensorflow==1.15.4
ERROR: No matching distribution found for tensorflow==1.15.4
any idea to solve this?
Thank you

**Comment by JackWBoynton** [comment]
You need to use python version 3.7 or lower to use tensorflow < 2.0

**Comment by Bturan19** [comment]
@TekpreXyz You may need to upgrade pip. Also, it'd be better to install all dependencies in a virtual environment and register the venv to jupyter.

**Comment by TekpreXyz** [comment]
> You need to use python version 3.7 or lower to use tensorflow < 2.0

thank you! i will try

**Comment by TekpreXyz** [comment]
> @TekpreXyz You may need to upgrade pip. Also, it'd be better to install all dependencies in a virtual environment and register the venv to jupyter.

sure, i already work on venv, thank you

**Comment by kfrancoi** [comment]
> I tried to create a lekage feature, to see if the model is able to exploit this feature and make some amazing profit. But even though, I leaked the future information as a feature, model still behaves like buy and hold.

Did you end up understanding why your model wasn't learning anything, even with some leakage features ?  I'm running into the same problem. During training, the loss function is not getting better, it's stays flat.

**Comment by Bturan19** [comment]
> > I tried to create a lekage feature, to see if the model is able to exploit this feature and make some amazing profit. But even though, I leaked the future information as a feature, model still behaves like buy and hold.
> 
> Did you end up understanding why your model wasn't learning anything, even with some leakage features ? I'm running into the same problem. During training, the loss function is not getting better, it's stays flat.

No, I couldnt understand.
